rosidl_typesupport_introspection_tests/test/test_typesupport_c.cpp

```diff
@@ -0,0 +1,2215 @@
+    {"MultiNested", 9},
+    {"Nested", 1},
+    {"Strings", 12},
+    {"UnboundedSequences", 32},
+    {"WStrings", 7},
+  };
+  const std::vector<std::string> services = {
+    "Arrays",
+    "BasicTypes",
+    "Empty",
+  };
+  // TODO(gbiggs): This should be const
+  std::unordered_map<std::string, uint32_t> services_message_member_counts = {
+    {"Arrays_Request", 31},
+    {"Arrays_Response", 31},
+    {"BasicTypes_Request", 14},
+    {"BasicTypes_Response", 14},
+    {"Empty_Request", 1},  // Not zero because a structure needs at least one member
+    {"Empty_Response", 1},  // Not zero because a structure needs at least one member
+  };
+  const std::vector<std::string> actions = {
+    "Fibonacci",
+    "NestedMessage",
+  };
+  // TODO(gbiggs): This should be const
+  std::unordered_map<std::string, uint32_t> action_message_member_counts = {
+    {"Fibonacci_Feedback", 1},
+    {"Fibonacci_FeedbackMessage", 2},
+    {"Fibonacci_GetResult_Request", 1},
+    {"Fibonacci_GetResult_Response", 2},
+    {"Fibonacci_Goal", 1},
+    {"Fibonacci_Result", 1},
+    {"Fibonacci_SendGoal_Request", 2},
+    {"Fibonacci_SendGoal_Response", 2},
+    {"NestedMessage_Feedback", 3},
+    {"NestedMessage_FeedbackMessage", 2},
+    {"NestedMessage_GetResult_Request", 1},
+    {"NestedMessage_GetResult_Response", 2},
+    {"NestedMessage_Goal", 3},
+    {"NestedMessage_Result", 3},
+    {"NestedMessage_SendGoal_Request", 2},
+    {"NestedMessage_SendGoal_Response", 2},
+  };
+
+  std::vector<std::string> service_messages;
+  std::vector<std::string> action_messages;
+  std::vector<std::string> action_services;
+  std::vector<std::string> all_messages;
+  std::vector<std::string> all_services;
+
+  std::unordered_map<std::string, uint32_t> all_messages_member_counts;
+};
+
+void get_symbol_from_library_with_checks(
+  void * library,  // in
+  const std::string & symbol_name,  // in
+  void ** symbol)  // out if not nullptr
+{
+  // Clear any previous error
+  dlerror();
+
+  // Look up the symbol's address
+  void * tmp_symbol = dlsym(library, symbol_name.c_str());
+  EXPECT_NE(tmp_symbol, nullptr);
+  if (symbol) {
+    *symbol = tmp_symbol;
+  }
+
+  // The correct way to determine if the symbol was found is to confirm that there was no error
+  // string set, i.e. dlerror() returns nullptr. This should be fatal to prevent a calling test
+  // from continuing with a null symbol, but we can't use ASSERT_* in a function that doesn't
+  // return void.
+  char * error_str = dlerror();
+  ASSERT_EQ(error_str, nullptr) << "Error finding symbol '" << symbol_name << "': " << error_str;
+}
+
+typedef const rosidl_message_type_support_t * (*get_message_ts_function)();
+typedef const rosidl_service_type_support_t * (*get_service_ts_function)();
+
+std::string make_message_typesupport_function_name(
+  std::string ns,
+  std::string type,
+  std::string interface)
+{
+  return "rosidl_typesupport_introspection_c__get_message_type_support_handle__" +
```

Comment:
@gbiggs hmm, we should probably use [rosidl_typesupport_interface](https://github.com/ros2/rosidl/blob/67bcca666d818823aba8ea08c12160a9badf4e41/rosidl_typesupport_interface/include/rosidl_typesupport_interface/macros.h) macros here. That won't do in macro form. Maybe it is a good time to add function equivalents.

Reply:
Totally agree.